libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -398,11 +401,238 @@ void AC_Avoid::adjust_velocity_polygon_fence(float kP, float accel_cmss, Vector2
+                if (stopping_point_dist <= (radius_cm - margin_cm)) {
+                    // stopping before before fence so no need to adjust for this circle
+                    continue;
+                }
+                // unsafe desired velocity - will not be able to stop before reaching margin from fence
+                // project stopping point radially onto fence boundary
+                // adjusted velocity will point towards this projected point at a safe speed
+                const Vector2f target_offset = stopping_point * ((radius_cm - margin_cm) / stopping_point_dist);
+                const Vector2f target_direction = target_offset - position_NE_rel;
+                const float distance_to_target = target_direction.length();
+                const float max_speed = get_max_speed(kP, accel_cmss, distance_to_target, dt);
+                desired_vel_cms = target_direction * (MIN(desired_speed,max_speed) / distance_to_target);
+            } else {
+                // implement stopping behaviour
+                // calculate stopping point plus a margin so we look forward far enough to intersect with circular fence
+                const Vector2f stopping_point_plus_margin = position_NE_rel + desired_vel_cms*((2.0f + margin_cm + get_stopping_distance(kP, accel_cmss, desired_speed))/desired_speed);
+                const float stopping_point_plus_margin_dist_cm = stopping_point_plus_margin.length();
+                const float dist_cm = safe_sqrt(dist_sq_cm);
+                if (dist_cm >= radius_cm - margin_cm) {
+                    // if vehicle has already breached margin around fence
+                    // if stopping point is even further from center (i.e. in wrong direction) then adjust speed to zero
+                    // otherwise user is backing away from fence so do not apply limits
+                    if (stopping_point_plus_margin_dist_cm >= dist_cm) {
+                        desired_vel_cms.zero();
+                    }
+                } else {
+                    // shorten vector without adjusting its direction
+                    Vector2f intersection;
+                    if (Vector2f::circle_segment_intersection(position_NE_rel, stopping_point_plus_margin, Vector2f(0.0f,0.0f), radius_cm - margin_cm, intersection)) {
+                        const float distance_to_target = MAX((intersection - position_NE_rel).length() - margin_cm, 0.0f);
+                        const float max_speed = get_max_speed(kP, accel_cmss, distance_to_target, dt);
+                        if (max_speed < desired_speed) {
+                            desired_vel_cms *= MAX(max_speed, 0.0f) / desired_speed;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+/*
+ * Adjusts the desired velocity for the exclusion circles
+ */
+void AC_Avoid::adjust_velocity_exclusion_circles(float kP, float accel_cmss, Vector2f &desired_vel_cms, float dt)
```

Comment:
it will be difficult to combine the inclusion and exclusion circle code because of the difference in slide behaviour but I think adjust_velocity_circle_fence() and adjust_velocity_inclusion_circles() could be combined.  I'm not going to do that for now unless people are really really keen because I'm not sure it's worth my time and effort at this moment because it will slowdown other developments we need to complete before we can start beta testing.